staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -6555,10 +6556,14 @@ type ObjectReference struct {
 // +structType=atomic
 type LocalObjectReference struct {
 	// Name of the referent.
-	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// This field is effectively required, but due to backwards compatibility is
+	// allowed to be empty. Instances of this type with an empty value here are
+	// almost certainly wrong.
 	// TODO: Add other useful fields. apiVersion, kind, uid?
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
 	// +optional
-	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// +default=""
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
```

Comment:
Is dropping `omitempty` required? (I can believe the tooling wouldn't allow setting a default and omitempty, because that's kind of crazy, but just wanted to make sure we were being forced to). Ideally we would just make changes that modify openapi, not serialization.

Dropping `omitempty` changes serialization of existing persisted data that accepted an empty value here. Normally that wouldn't be a problem but we know we have some components that are very sensitive to that (like the kubelet hashing serialized json content to decide if a container changed).

*If* we have to drop omitempty, we need to enumerate the places this was used, which of those permitted empty name values in validation, and make sure none of those places will cause issues if the serialization of existing persisted data changes to include `name: ""` on an API server upgrade.

Reply:
ISTR, but not 100% that omitempty was used to indicate optionalness in some contexts.  @pmalek I hate throwing more and more on this PR, but maybe you can set it up as one commit with the tags, then run your local test to load CRD with PodSpec and report back?

Reply:
the openapi generator falls back to looking at `omitempty` as a measure of optionality if neither +optional nor +required is present:

https://github.com/kubernetes/kubernetes/blob/e342ab05bb903519350cedac784898529eaef06b/vendor/k8s.io/kube-openapi/pkg/generators/openapi.go#L88-L103

For this field, it actually *is* optional from a client perspective (they're not required to send it in the request)

For the IP field, it is *not* optional, but the explicit `+required` annotation *should* cover that for us

Reply:
> @pmalek I hate throwing more and more on this PR

I'm also sorry for picking at this while you're trying to help us undo a bug we caused, thanks for the help. Just looking to make the change as targeted and side-effect free as possible.

Reply:
No worries about this taking longer, better to get this right than rush and fail to fix it.

---

My tests might be a bit limited but since I have not figured out how to easily incorporate changes from this PR into my CRD generation with kubebuilder I'm basically using the definitions of structs from this PR and embedding that into my CRD instead of the `PodTemplateSpec`.

Trying this out with 

```
type MyCRD struct {
...
	// ImagePullSecrets is an optional list of references to secrets in the same namespace to use for pulling any of the images used by this PodSpec.
	// If specified, these secrets will be passed to individual puller implementations for them to use.
	// More info: https://kubernetes.io/docs/concepts/containers/images#specifying-imagepullsecrets-on-a-pod
	// +optional
	// +patchMergeKey=name
	// +patchStrategy=merge
	// +listType=map
	// +listMapKey=name
	ImagePullSecrets []LocalObjectReference `json:"imagePullSecrets,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,15,rep,name=imagePullSecrets"`
}

// LocalObjectReference contains enough information to let you locate the
// referenced object inside the same namespace.
// +structType=atomic
type LocalObjectReference struct {
	// Name of the referent.
	// This field is effectively required, but due to backwards compatibility is
	// allowed to be empty. Instances of this type with an empty value here are
	// almost certainly wrong.
	// TODO: Add other useful fields. apiVersion, kind, uid?
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
	// +optional
	// +kubebuilder:default=""
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}
```

It does work for me. Do note that I've used `+kubebuilder:default=""` instead of the `+default=""` because `kubebuilder` didn't want to pick that one up.

`omitempty` or not it allows me to apply the CRD.

Reply:
> Do note that I've used +kubebuilder:default="" instead of the +default="" because kubebuilder didn't want to pick that one up.

Hmm... I don't love putting kube-builder-specific annotations here ... @jpbetz @alexzielenski I *really* thought we taught kubebuilder about standard kubernetes `+default` annotations ... did I hallucinate that?

Reply:
I'm just saying that in my tests where I generate CRDs via `controller-gen` like so:

```
controller-gen paths="./api/...;" "+crd:generateEmbeddedObjectMeta=true" +output:crd:artifacts:config=config/crd/bases
```

`+kubebuilder:default=""` resulted in default being added while `+default=""` did not.

I'm using `0.14.0` version of controller-gen.

Reply:
After setting `+default`, did you re-run the codegen and openapispec?

Reply:
I'm trying to do my own testing, but I still can't repro.

Given `gateway-operator` at hash f047d, I ran `make install` and got:

```
cd config/components/manager-image/ && /tmp/gateway-operator/bin/installs/kustomize/5.3.0/bin/kustomize edit set image docker.io/kong/gateway-operator-oss=docker.io/kong/gateway-operator-oss:1.2.3
/tmp/gateway-operator/bin/installs/kube-controller-tools/0.14.0/bin/controller-gen paths="./pkg/utils/kubernetes/resources/clusterroles/; ./pkg/utils/kubernetes/reduce/; ./controller/...; ./api/...;" rbac:roleName=manager-role output:rbac:dir=config/rbac/role
/tmp/gateway-operator/bin/installs/kube-controller-tools/0.14.0/bin/controller-gen paths="./pkg/utils/kubernetes/resources/clusterroles/; ./pkg/utils/kubernetes/reduce/; ./controller/...; ./api/...;" webhook
/tmp/gateway-operator/bin/installs/kube-controller-tools/0.14.0/bin/controller-gen paths="./pkg/utils/kubernetes/resources/clusterroles/; ./pkg/utils/kubernetes/reduce/; ./controller/...; ./api/...;" "+crd:generateEmbeddedObjectMeta=true" +output:crd:artifacts:config=config/crd/bases
/tmp/gateway-operator/bin/installs/kustomize/5.3.0/bin/kustomize build /home/thockin/go/pkg/mod/sigs.k8s.io/gateway-api@v1.0.0/config/crd | kubectl apply -f -
customresourcedefinition.apiextensions.k8s.io/gatewayclasses.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/gateways.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/httproutes.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/referencegrants.gateway.networking.k8s.io configured
/tmp/gateway-operator/bin/installs/kustomize/5.3.0/bin/kustomize build config/crd | kubectl apply --server-side -f -
customresourcedefinition.apiextensions.k8s.io/controlplanes.gateway-operator.konghq.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/dataplanemetricsextensions.gateway-operator.konghq.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/dataplanes.gateway-operator.konghq.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/gatewayconfigurations.gateway-operator.konghq.com serverside-applied
```

I am unclear how to make it FAIL so I can poke at solution paths.  What am I missing?

Reply:
Sorry for not providing clear repro instructions. This repo didn't merge the change to 1.30 k8s libs because of the bug that we're trying to fix here.

The `dependabot/go_modules/k8s.io/kubernetes-1.30.0` branch or this PR https://github.com/Kong/gateway-operator/pull/107 contain the "broken" version. If you check that out and run 

```
make manifests  && kubectl kustomize ./config/crd | kubectl apply --server-side -f -
```

You will get the original error.

Reply:
OK.  I have a repro now.  Thanks

Reply:
I added a go.mod replace directive using my local kubernetes tree which is roughly HEAD.  Without changing anything: 

```
Error from server (Invalid): CustomResourceDefinition.apiextensions.k8s.io "gatewayconfigurations.gateway-operator.konghq.com" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[hostAliases].items.properties[ip].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[hostAliases].items.properties[ip].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property]

```

I added `// +required` on the HostAlias.IP field.  Now I get:

```
Error from server (Invalid): CustomResourceDefinition.apiextensions.k8s.io "gatewayconfigurations.gateway-operator.konghq.com" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[hostAliases].items.properties[ip].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[hostAliases].items.properties[ip].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property]
```

So that's not sufficient.  I took out `omitempty`:

```
Error from server (Invalid): CustomResourceDefinition.apiextensions.k8s.io "gatewayconfigurations.gateway-operator.konghq.com" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property]
```

So that works for HostAlias.IP.  For giggles I did put omitempty back and added `// +default=""` and that did NOT work.

Moving on, I added `// +default=""` to LocalObjectReference.Name:

```
Error from server (Invalid): CustomResourceDefinition.apiextensions.k8s.io "gatewayconfigurations.gateway-operator.konghq.com" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[dataPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property, spec.validation.openAPIV3Schema.properties[spec].properties[controlPlaneOptions].properties[deployment].properties[podTemplateSpec].properties[spec].properties[imagePullSecrets].items.properties[name].default: Required value: this property is in x-kubernetes-list-map-keys, so it must have a default or be a required property]
```

I added `// +kubebuilder:default=""` and it worked.

For kicks I tried `// +kubebuilder:required` with omitempty (for HostAlias) and it did not work.  

 @jpbetz @alexzielenski for expertise

Reply:
This uses controller-gen v0.14.0.  The latest is v0.15.0, but release notes don't indicate anything related to this.

Reply:
Here's the diff, minus codegen, that seems to work:

```
diff --git a/staging/src/k8s.io/api/core/v1/types.go b/staging/src/k8s.io/api/core/v1/types.go
index b7b7640bd70..526c537988d 100644
--- a/staging/src/k8s.io/api/core/v1/types.go
+++ b/staging/src/k8s.io/api/core/v1/types.go
@@ -4115,7 +4115,8 @@ const (
 // pod's hosts file.
 type HostAlias struct {
        // IP address of the host file entry.
-       IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
+       // +required
+       IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
        // Hostnames for the above IP address.
        // +listType=atomic
        Hostnames []string `json:"hostnames,omitempty" protobuf:"bytes,2,rep,name=hostnames"`
@@ -6557,9 +6558,15 @@ type ObjectReference struct {
 // +structType=atomic
 type LocalObjectReference struct {
        // Name of the referent.
+       // This field is effectively required, but due to backwards compatibility is
+       // allowed to be empty. Instances of this type with an empty value here are
+       // almost certainly wrong.
        // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
        // TODO: Add other useful fields. apiVersion, kind, uid?
        // +optional
+       // +default=""
+       // +kubebuilder:default=""
+       // TODO: Drop `kubebuilder:default` when controller-gen doesn't need it
        Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
 }
 
```

Reply:
I don't love putting kubebuilder annotations in, but can live with it for now... can we file an issue to teach kube-builder about `+default` as a follow-up? I *really* thought that had already been done :-/

Reply:
> This uses controller-gen v0.14.0. The latest is v0.15.0, but release notes don't indicate anything related to this.

I've checked with 0.15 and same results.

@thockin Your diff aligns with what I pushed in https://github.com/kubernetes/kubernetes/pull/124553/files#diff-5ffbebb40a29f8e491f4ff1e07c618eb3edcc0d5f00a1a71445272d6767d7883 minus the kubebuilder annotation. For some reason I thought that this would work differently when merged so I incorrectly assumed that we didn't need the `kubebuilder` comment. 


---


I've created an issue for kubebuilder https://github.com/kubernetes-sigs/kubebuilder/issues/3896 (we can always transfer to controller-tools if it should land there, I'm not sure ATM).

Reply:
@pmalek - it's your PR, so once you incorporate all the things we have discovered here and sufficient comments, ping for re-review :)